test/integration/scheduler/preemption/preemption_test.go

```diff
@@ -452,59 +463,561 @@ func TestPreemption(t *testing.T) {
+				scheduler.WithPodInitialBackoffSeconds(0),
+			)
+			testutils.SyncSchedulerInformerFactory(testCtx)
+			cs := testCtx.ClientSet
+
+			if preemptionPlugin == nil {
+				t.Fatalf("the preemption plugin should be initialized")
+			}
+
+			logger, _ := ktesting.NewTestContext(t)
+			featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.SchedulerAsyncPreemption, true)
+
+			createdPods := []*v1.Pod{}
+			defer testutils.CleanupPods(testCtx.Ctx, cs, t, createdPods)
+
+			ctx, cancel := context.WithCancel(context.Background())
+			defer cancel()
+
+			if _, err := cs.CoreV1().Nodes().Create(ctx, node, metav1.CreateOptions{}); err != nil {
+				t.Fatalf("Failed to create an initial Node %q: %v", node.Name, err)
+			}
+			defer func() {
+				if err := cs.CoreV1().Nodes().Delete(ctx, node.Name, metav1.DeleteOptions{}); err != nil {
+					t.Fatalf("Failed to delete the Node %q: %v", node.Name, err)
+				}
+			}()
+
+			for _, scenario := range test.scenarios {
+				t.Logf("Running scenario: %s", scenario.name)
+				switch {
+				case scenario.createPod != nil:
+					if scenario.createPod.count == nil {
+						scenario.createPod.count = ptr.To(1)
 					}
-					pod, err := cs.CoreV1().Pods(p.Namespace).Get(testCtx.Ctx, p.Name, metav1.GetOptions{})
-					if err != nil {
-						t.Errorf("Error %v when getting the updated status for pod %v/%v ", err, p.Namespace, p.Name)
+
+					for i := 0; i < *scenario.createPod.count; i++ {
+						pod, err := cs.CoreV1().Pods(testCtx.NS.Name).Create(ctx, scenario.createPod.pod, metav1.CreateOptions{})
+						if err != nil {
+							t.Fatalf("Failed to create a Pod %q: %v", pod.Name, err)
+						}
+						createdPods = append(createdPods, pod)
 					}
-					_, cond := podutil.GetPodCondition(&pod.Status, v1.DisruptionTarget)
-					if cond == nil {
-						t.Errorf("Pod %q does not have the expected condition: %q", klog.KObj(pod), v1.DisruptionTarget)
+				case scenario.schedulePod != nil:
+					if err := wait.PollUntilContextTimeout(testCtx.Ctx, time.Millisecond*200, wait.ForeverTestTimeout, false, func(ctx context.Context) (bool, error) {
+						activePods := testCtx.Scheduler.SchedulingQueue.PodsInActiveQ()
+						return len(activePods) != 0, nil
+					}); err != nil {
+						t.Fatalf("Expected the pod %s to be scheduled, but no pod arrives at the activeQ", scenario.schedulePod.podName)
 					}
-				} else {
-					if p.DeletionTimestamp != nil {
-						t.Errorf("Didn't expect pod %v to get preempted.", p.Name)
+
+					if err := wait.PollUntilContextTimeout(testCtx.Ctx, time.Millisecond*200, wait.ForeverTestTimeout, false, func(ctx context.Context) (bool, error) {
+						return testCtx.Scheduler.SchedulingQueue.PodsInActiveQ()[0].Name == scenario.schedulePod.podName, nil
+					}); err != nil {
+						t.Fatalf("The pod %s is expected to be scheduled, but the top Pod is %s", scenario.schedulePod.podName, testCtx.Scheduler.SchedulingQueue.PodsInActiveQ()[0].Name)
+					}
```

Comment:
Do we need to poll twice here?

Reply:
Good catch, I splitted one into two when debugging the problem. We can bring back them into one.